pex_reactor.go

```diff
@@ -96,86 +111,97 @@ func (pexR *PEXReactor) Receive(chID byte, src *Peer, msgBytes []byte) {
 		for _, addr := range msg.Addrs {
-			pexR.book.AddAddress(addr, srcAddr)
+			r.book.AddAddress(addr, srcAddr)
 		}
 	default:
 		log.Warn(Fmt("Unknown message type %v", reflect.TypeOf(msg)))
 	}
 
 }
 
-// Asks peer for more addresses.
-func (pexR *PEXReactor) RequestPEX(peer *Peer) {
-	peer.Send(PexChannel, struct{ PexMessage }{&pexRequestMessage{}})
+// RequestPEX asks peer for more addresses.
+func (r *PEXReactor) RequestPEX(p *Peer) {
+	p.Send(PexChannel, struct{ PexMessage }{&pexRequestMessage{}})
+}
+
+// SendAddrs sends addrs to the peer.
+func (r *PEXReactor) SendAddrs(p *Peer, addrs []*NetAddress) {
+	p.Send(PexChannel, struct{ PexMessage }{&pexAddrsMessage{Addrs: addrs}})
+}
+
+// SetEnsurePeersPeriod sets period to ensure peers connected.
+func (r *PEXReactor) SetEnsurePeersPeriod(d time.Duration) {
+	r.ensurePeersPeriod = d
+}
+
+// SetMaxMsgCountByPeer sets maximum messages one peer can send to us during 'msgCountByPeerFlushInterval'.
+func (r *PEXReactor) SetMaxMsgCountByPeer(v uint16) {
+	r.maxMsgCountByPeer = v
 }
 
-func (pexR *PEXReactor) SendAddrs(peer *Peer, addrs []*NetAddress) {
-	peer.Send(PexChannel, struct{ PexMessage }{&pexAddrsMessage{Addrs: addrs}})
+// ReachedMaxMsgCountForPeer returns true if we received too many
+// messages from peer with address `addr`.
+func (r *PEXReactor) ReachedMaxMsgCountForPeer(addr string) bool {
+	return r.msgCountByPeer[addr] >= r.maxMsgCountByPeer
 }
 
 // Ensures that sufficient peers are connected. (continuous)
-func (pexR *PEXReactor) ensurePeersRoutine() {
+func (r *PEXReactor) ensurePeersRoutine() {
 	// Randomize when routine starts
-	time.Sleep(time.Duration(rand.Int63n(500*ensurePeersPeriodSeconds)) * time.Millisecond)
+	ensurePeersPeriodMs := r.ensurePeersPeriod.Nanoseconds() / 1e6
+	time.Sleep(time.Duration(rand.Int63n(ensurePeersPeriodMs)) * time.Millisecond)
 
 	// fire once immediately.
-	pexR.ensurePeers()
+	r.ensurePeers()
+
 	// fire periodically
-	timer := NewRepeatTimer("pex", ensurePeersPeriodSeconds*time.Second)
-FOR_LOOP:
+	ticker := time.NewTicker(r.ensurePeersPeriod)
+
 	for {
 		select {
-		case <-timer.Ch:
-			pexR.ensurePeers()
-		case <-pexR.Quit:
-			break FOR_LOOP
+		case <-ticker.C:
+			r.ensurePeers()
+		case <-r.Quit:
+			ticker.Stop()
+			return
 		}
 	}
-
-	// Cleanup
-	timer.Stop()
 }
 
-// Ensures that sufficient peers are connected. (once)
-func (pexR *PEXReactor) ensurePeers() {
-	numOutPeers, _, numDialing := pexR.Switch.NumPeers()
+// ensurePeers ensures that sufficient peers are connected. (once)
+func (r *PEXReactor) ensurePeers() {
+	numOutPeers, _, numDialing := r.Switch.NumPeers()
 	numToDial := minNumOutboundPeers - (numOutPeers + numDialing)
 	log.Info("Ensure peers", "numOutPeers", numOutPeers, "numDialing", numDialing, "numToDial", numToDial)
 	if numToDial <= 0 {
 		return
 	}
-	toDial := NewCMap()
+
+	toDial := make(map[string]*NetAddress)
 
 	// Try to pick numToDial addresses to dial.
-	// TODO: improve logic.
```

Comment:
If we mark the peer good after successful connection and pick addresses with the bias = 100% (always from new buckets), we don't need to check `alreadyConnected` (since all the peers with whom we already have a connection will be in the old buckets) https://github.com/tendermint/go-p2p/pull/10/commits/b898bc3703f6fb61f413174412ffb7990972fdf4